src/wav_output.rs

```diff
@@ -2,35 +2,68 @@ use crate::common::assert_error_traits;
 use crate::Source;
 use hound::{SampleFormat, WavSpec};
 use std::path;
+use std::io;
 use std::sync::Arc;
 
 #[derive(Debug, thiserror::Error, Clone)]
 pub enum ToWavError {
-    #[error("Could not create wav file")]
+    #[error("Opening file for writing")]
+    OpenFile(#[source] Arc<std::io::Error>),
+    #[error("Could not create wav writer")]
     Creating(#[source] Arc<hound::Error>),
-    #[error("Failed to write samples to wav file")]
+    #[error("Failed to write samples writer")]
     Writing(#[source] Arc<hound::Error>),
     #[error("Failed to update the wav header")]
     Finishing(#[source] Arc<hound::Error>),
 }
 assert_error_traits!(ToWavError);
 
-/// This procedure saves Source's output into a wav file. The output samples format is 32-bit float.
-/// This function is intended primarily for testing and diagnostics. It can be used to see
+/// Saves Source's output into a wav file. The output samples format is 32-bit
+/// float. This function is intended primarily for testing and diagnostics. It can be used to see
 /// the output without opening output stream to a real audio device.
 ///
 /// If the file already exists it will be overwritten.
 pub fn output_to_wav(
-    source: &mut impl Source,
+    source: impl Source,
     wav_file: impl AsRef<path::Path>,
+) -> Result<(), ToWavError> {
+    let mut file = std::fs::File::create(wav_file)
+        .map_err(Arc::new)
+        .map_err(ToWavError::OpenFile)?;
+    collect_to_wav(source, &mut file)
+}
+
+/// Saves Source's output into a writer. The output samples format is 32-bit float. This function
+/// is intended primarily for testing and diagnostics. It can be used to see the output without
+/// opening output stream to a real audio device.
+///
+/// # Example
+/// ```rust
+/// # use rodio::static_buffer::StaticSamplesBuffer;
+/// # use rodio::collect_to_wav;
+/// # const SAMPLES: [rodio::Sample; 5] = [0.0, 1.0, 2.0, 3.0, 4.0];
+/// # let source = StaticSamplesBuffer::new(
+/// #     1.try_into().unwrap(), 
+/// #     1.try_into().unwrap(), 
+/// #     &SAMPLES
+/// # );
+/// let mut writer = std::io::Cursor::new(Vec::new());
+/// collect_to_wav(source, &mut writer)?;
+/// let wav_bytes: Vec<u8> = writer.into_inner();
+/// # Ok::<(), Box<dyn std::error::Error>>(())
+/// ```
+pub fn collect_to_wav(
+    source: impl Source,
+    writer: &mut (impl io::Write + io::Seek),
 ) -> Result<(), ToWavError> {
     let format = WavSpec {
         channels: source.channels().get(),
         sample_rate: source.sample_rate().get(),
         bits_per_sample: 32,
         sample_format: SampleFormat::Float,
     };
-    let mut writer = hound::WavWriter::create(wav_file, format)
+    let writer = io::BufWriter::new(writer);
```

Comment:
I've never really worked with `BufWriter` before, but I believe there are some caveats to dropping and flushing:

> It is critical to call [flush](https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.flush) before BufWriter<W> is dropped.

Reply:
I did not realize that, makes sense. Good catch though!